sound/soc/sof/intel/hda-pcm.c

```diff
@@ -148,13 +148,21 @@ snd_pcm_uframes_t hda_dsp_pcm_pointer(struct snd_sof_dev *sdev,
 				      struct snd_pcm_substream *substream)
 {
 	struct snd_soc_pcm_runtime *rtd = substream->private_data;
-	struct snd_soc_component *scomp = sdev->sof_audio_data->component;
 	struct hdac_stream *hstream = substream->runtime->private_data;
 	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
-	struct snd_sof_pcm *spcm;
+	struct snd_sof_client *client;
+	struct snd_sof_pcm *spcm = NULL;
 	snd_pcm_uframes_t pos;
 
-	spcm = snd_sof_find_spcm_dai(scomp, rtd);
+	/* search all clients to find the spcm with matching dai id */
+	list_for_each_entry(client, &sdev->client_list, list) {
+		if (client->type != SOF_CLIENT_AUDIO)
+			continue;
+
+		spcm = snd_sof_find_spcm_dai(&client->pdev->dev, rtd);
+		if (spcm)
+			break;
+	}
```

Comment:
the pointer callback is quite latency sensitive, we should probably find a means to match dai id and spcm on startup and avoid this search every period (when IPC updates are requested)

Reply:
@plbossart yes I could do that to reduce the recursive search here but we'd still have one loop to find the spcm matching the dai id from the list of open streams in case there are more than one. Is that ok?

Reply:
Sorry I don't get the answer. I suggested this loop be removed and you replied that it can be removed but is still needed. what I am missing?

Reply:
@plbossart there are 2 loops here. The outer loops for the list of clients and the inner loop in snd_sof_find_spcm_dai(). The inner loop has always been there, its only the outer loop that is an addition in this PR because now we have multiple clients, we have to search them all to find the spcm matcing the dai ID. 

Maybe I didnt understand your suggestion correctly. Are you suggesting to remove just the outer loop to iterate through the client list? or both the loops? In either case, I dont think we can remove the loop entirely because even if saved the spcm at startup, in the case of multiple streams open, we'd need to find one which one to choose here.